In [4]:
!pip install easyocr opencv-python-headless

In [10]:
import easyocr
import cv2
import os

reader = easyocr.Reader(['en'], gpu=True)

meme_files = ['meme1.png', 'meme2.png', 'meme3.png']

print("--- Final Assignment Submission Output --- \n")

for image_path in meme_files:
    if os.path.exists(image_path):
        # We use allowlist to keep the output clean of random symbols
        # contrast_ths=0.05 makes the AI more sensitive to low-contrast text
        results = reader.readtext(image_path,
                                  detail=0,
                                  paragraph=True,
                                  contrast_ths=0.05,
                                  allowlist='0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ ')

        extracted_text = " ".join(results).strip()

        print(f"Image: {image_path}")
        print(f"Extracted Text: {extracted_text}")
        print("-" * 40)

print("\n--- Extraction Complete ---")

--- Final Assignment Submission Output --- 

Image: meme1.png
Extracted Text: Going to LHC Sleeping in bed Made with Pinata Farms
----------------------------------------
Image: meme2.png
Extracted Text: SPEXIT HHOURS DEBUCGIEG CODE TT FINAQLY RUNS WHtHOUT ERRORS
----------------------------------------
Image: meme3.png
Extracted Text: tHaTs A gReAt IdEa 0  2 9
----------------------------------------

--- Extraction Complete ---


In [12]:
import easyocr
from PIL import Image, ImageDraw
import os

reader = easyocr.Reader(['en'], gpu=True)


meme_files = ['meme1.png', 'meme2.png', 'meme3.png']

print("--- Assignment 2.2: The Confidence Filter --- \n")

for image_path in meme_files:
    if os.path.exists(image_path):

        img = Image.open(image_path).convert("RGB")
        draw = ImageDraw.Draw(img)

        # detail=1 returns (bounding box, text, confidence)
        results = reader.readtext(image_path, detail=1)

        print(f"Results for {image_path}:")
        final_output_text = []

        for (bbox, text, prob) in results:
            # 2. Check the Confidence Score (prob)
            if prob > 0.5:
                final_output_text.append(text)
                print(f"  [CONFIDENT] {text} (Score: {prob:.2f})")


                # bbox is [[TL], [TR], [BR], [BL]]
                points = [tuple(p) for p in bbox]
                draw.polygon(points, outline="red", width=3)
            else:
                print(f"  [REJECTED]  {text} (Score: {prob:.2f})")


        print(f"\nExtracted Text: {' '.join(final_output_text)}")

        img.save(f"boxed_{image_path}")
        print(f"Visualization saved as boxed_{image_path}")
        print("-" * 45)

print("\n--- Task Complete ---")

--- Assignment 2.2: The Confidence Filter --- 

Results for meme1.png:
  [CONFIDENT] Going to LHC (Score: 0.97)
  [CONFIDENT] Sleeping in bed (Score: 0.92)
  [CONFIDENT] Made with Pinata Farms (Score: 0.97)

Extracted Text: Going to LHC Sleeping in bed Made with Pinata Farms
Visualization saved as boxed_meme1.png
---------------------------------------------
Results for meme2.png:
  [REJECTED]  SPEXIT HHOURS DEBUCGIEG CODE (Score: 0.15)
  [REJECTED]  [T FINAQLY RUNS WHtHOUT ERRORS (Score: 0.24)

Extracted Text: 
Visualization saved as boxed_meme2.png
---------------------------------------------
Results for meme3.png:
  [CONFIDENT] tHaTs A gReAt IdEa (Score: 0.73)
  [REJECTED]  @ } (Score: 0.03)
  [CONFIDENT] 2 (Score: 0.95)
  [REJECTED]  ) (Score: 0.23)

Extracted Text: tHaTs A gReAt IdEa 2
Visualization saved as boxed_meme3.png
---------------------------------------------

--- Task Complete ---


Assignment 2.1 & 2.2 OCR Extraction Results

Image: meme1.png
Extracted Text: Going to LHC Sleeping in bed Made with Pinata Farms
Confidence: High (Avg > 0.90)

Image: meme2.png
Extracted Text: [REJECTED - Confidence below 0.5]
Reason: Image noise and stylized font resulted in scores of 0.15 and 0.24.

Image: meme3.png
Extracted Text: tHaTs A gReAt IdEa 2
Confidence: Mixed (0.73 - 0.95)
Note: Random symbols (@, }, )) were successfully filtered out.